### ISSUE TITLE
The Endpoint of OpenAI model is not allow to change, how should I add 3rd party LLMs?

### ISSUE THREAD

**Issue by freemank1224** [microsoft/data-formulator#49]
I notice that I can modify some files in the src folder, but this still can't work when I use data_formulator. So can you provide the users the oppertunity to use 3rd party endpoints ?

**Comment by Chenglong-MS** [comment]
In order to be able to modify source code and reflect it in data formulator, you should follow the development build (instead of using `pip install data_formulator`).

You can check out the development build here https://github.com/microsoft/data-formulator/blob/main/DEVELOPMENT.md

You'll need to complete both backend build and frontend build, and visit http://localhost:3000/ to see the live version. 

Some tips: if you made updates to backend, you'll need to restart the backend server with `.\local_server.sh` to reflect changes to the custom data formulator you are developing.

**Comment by pmccabe157** [comment]
Where in the app can i change the endpoint e.g. to use an OLLAMA model?

**Comment by ramixpe** [comment]
i would love to see the Ollama as well as back end please

**Comment by Chenglong-MS** [comment]
Here is the place to update endpoint to use other models. https://github.com/microsoft/data-formulator/blob/main/py-src/data_formulator/agents/client_utils.py

There seems to be enough interests? We can update data formulator to support Ollama as part of the new version release.

**Comment by 82ndAirborneDiv** [comment]
I would love to see a way to use other LLM. And possibly change the API URL. And also to use it locally with Ollama.

**Comment by Chenglong-MS** [comment]
yes, this sounds like a very good approach. We'll test it out and do a PR soon.

**Comment by Chittaranjans** [comment]
OKay

**Comment by Chenglong-MS** [comment]
A work in progress supporting non-openai models is here: https://github.com/microsoft/data-formulator/tree/dev

I have tested with Ollama/OpenAI/AzureOpenAI, would do a bit more testing before doing a PR.

**Comment by Chittaranjans** [comment]
What i planned and tested , It work for all LLM , only the UI part for modelselection will have to do accordingly with the [client_utils.py](https://github.com/microsoft/data-formulator/blob/main/py-src/data_formulator/agents/client_utils.py).

**Comment by Chenglong-MS** [comment]
Checkout this PR: https://github.com/microsoft/data-formulator/pull/81, should work now! Will merge to main soon.

![Image](https://github.com/user-attachments/assets/ed2bad09-a7a0-4ede-a422-606f4bd7a4dd)

I have updated throughout (UI, agents, utils) so that Data Formulator can work with custom models.

So far by experience is that modes with good code generation and instruction following capabilities work best (gpt-4o, gpt-4o-mini, claude-3-5-sonnet etc). 

Small local models (llama3.2, qwen2.5-coder:3b, codellama:7b) tends to ignore instructions to generate code, and can fail frequently on data formulation steps.

**Comment by giyosphere** [comment]
You need to incorporate third-party LLMs without changing OpenAI’s fixed endpoint, you can use a middleware to route requests based on task requirements, call different LLMs selectively within your app, run local models if supported, or combine responses from multiple LLMs for richer output.

**Comment by Chenglong-MS** [comment]
Please check out the new release with custom model support: https://github.com/microsoft/data-formulator/releases/tag/0.1.5

@giyosphere made a good point. It would be good to batch some queries and adaptively choose model based on task requirement (different agent may have more suitable models). That would be a feature improvement.

**Comment by banalg** [comment]
Hello,

At my company, we use OpenWebUI, which provides a chatbot GUI and exposes LLM APIs in both OpenAI and Ollama formats. Additionally, it allows us to secure these endpoints with an API key.

However, I’ve encountered issues configuring MS DataFormulator with both the Ollama and OpenAI endpoints:

- The **Ollama endpoint** does not support setting an API key.
- The **OpenAI endpoint** does not allow changing the base URL.

Would it be possible to either:

- Enable API key support for the Ollama endpoint, or
- Allow customization of the base URL for the OpenAI endpoint?

This would greatly help in integrating MS DataFormulator with OpenWebUI.

Thank you for your consideration!

Best regards,

> Checkout this PR: [#81](https://github.com/microsoft/data-formulator/pull/81), should work now! Will merge to main soon.
> 
> ![Image](https://github.com/user-attachments/assets/ed2bad09-a7a0-4ede-a422-606f4bd7a4dd)
> 
> I have updated throughout (UI, agents, utils) so that Data Formulator can work with custom models.
> 
> So far by experience is that modes with good code generation and instruction following capabilities work best (gpt-4o, gpt-4o-mini, claude-3-5-sonnet etc).
> 
> Small local models (llama3.2, qwen2.5-coder:3b, codellama:7b) tends to ignore instructions to generate code, and can fail frequently on data formulation steps.
